pkg/metaclient/kvclient/etcdkv/etcd_client.go

```diff
@@ -0,0 +1,307 @@
+	"go.etcd.io/etcd/clientv3"
+	pb "go.etcd.io/etcd/etcdserver/etcdserverpb"
+	"go.uber.org/zap"
+)
+
+const (
+	DefaultRevisionChanSize = 128
+)
+
+// etcdClientImpl is the etcd implement of Client interface
+type etcdClientImpl struct {
+	cli *etcdImpl
+}
+
+func (c *etcdClientImpl) Close() {
+	c.cli.Close()
+}
+
+func NewEtcdClientImpl(c *etcdImpl) *etcdClientImpl {
+	return &etcdClientImpl{
+		cli: c,
+	}
+}
+
+// etcdImpl is the etcd implement of KVClient interface
+type etcdImpl struct {
+	cli           *clientv3.Client
+	compactCancel context.CancelFunc
+	revCh         chan int64
+	latestRevT    int64
+	closeMu       sync.Mutex
+	wg            *sync.WaitGroup
+}
+
+func startCompactRoutine(ctx context.Context, wg *sync.WaitGroup, ch <-chan int64, cli *clientv3.Client) {
+	defer wg.Done()
+	var rev, latestRev int64
+	var compactInterval bool
+	ticker := time.NewTicker(time.Second * 5)
+	for {
+		select {
+		case <-ctx.Done():
+			log.L().Info("receive context done, return")
+			return // cancel routine normally
+		case rrev, ok := <-ch:
+			if !ok {
+				log.L().Info("receive revision from channel fail, return")
+				return
+			}
+			rev = rrev
+		case <-ticker.C:
+			compactInterval = true
+		}
+
+		if compactInterval && rev > latestRev {
+			childCtx, cancel := context.WithTimeout(ctx, time.Second*3)
+			defer cancel()
+			_, err := cli.Compact(childCtx, rev)
+			if err != nil {
+				log.L().Warn("compact revision fail", zap.Int64("revision", rev), zap.Error(err))
+			}
+			compactInterval = false
+			latestRev = rev
+		}
+	}
+}
+
+func NewEtcdImpl(config *metaclient.Config) (metaclient.KVClient, error) {
+	conf := config.Clone()
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints: conf.Endpoints,
+		// [TODO] TLS
+		// Username: conf.Auth.Username,
+		// Password: conf.Auth.Password,
+		// [TODO] LOG
+	})
+	if err != nil {
+		return nil, cerrors.ErrMetaNewClientFail.Wrap(err)
+	}
+
+	var wg sync.WaitGroup
+	// start compact routine to avoid user space expeed
+	ctx, cancel := context.WithCancel(context.Background())
+	revCh := make(chan int64, DefaultRevisionChanSize)
+	wg.Add(1)
+	go startCompactRoutine(ctx, &wg, revCh, cli)
+
+	c := &etcdImpl{
+		cli:           cli,
+		compactCancel: cancel,
+		revCh:         revCh,
+		latestRevT:    time.Now().Unix(),
+		wg:            &wg,
+	}
+	return c, nil
+}
+
+func (c *etcdImpl) getLatestRev(h *pb.ResponseHeader) {
+	now := time.Now().Unix()
+	if (now - atomic.LoadInt64(&c.latestRevT)) > 1 {
+		atomic.StoreInt64(&c.latestRevT, now)
+		c.revCh <- h.Revision
+	}
+}
+
+func (c *etcdImpl) getEtcdOptions(op metaclient.Op) []clientv3.OpOption {
+	etcdOps := make([]clientv3.OpOption, 0, 3)
+	switch {
+	case op.IsOptsWithPrefix():
+		etcdOps = append(etcdOps, clientv3.WithPrefix())
+	case op.IsOptsWithFromKey():
+		etcdOps = append(etcdOps, clientv3.WithFromKey())
+	case op.IsOptsWithRange():
+		etcdOps = append(etcdOps, clientv3.WithRange(string(op.RangeBytes())))
+	}
```

Comment:
Do we need a `default` branch?

Reply:
Without a `default` branch seems ok here.